Policy for software package requests (#1387)
alan-turing-institute/data-safe-haven: docs/processes/software_package_approval.md

## Background

Given the safeguards afforded by the safe haven environment, and the separation of responsibilities between its constituent resources, the level of risk associated with the introduction of third party software packages is considered low.
Moreover, access to the environment is carefully controlled and there is a presumption of trust in the individual researchers to whom access is granted.

Nevertheless, the introduction of any software into the safe haven must be considered against the potential risks of:

- approved users having access to data to which they shouldn't (e.g. from data mixing)
- unapproved users having access to data (e.g. from a data breach)
- poisoning of data and/or outputs
- resource misuse (allocation of computational resources for unintended or wasteful purposes).

Such risks may originate unwittingly, from a user who wants to "just get the job done", or from a user, network team member or administrator acting maliciously.

Specific risks which this policy aims to mitigate include:

- package name squatting (allowlisting a similarly-named package instead of the intended one)

> - privilege escalation attacks (enabling a user to gain elevated access permissions).

Reviewer: It might be worth adding a point here about data ingress. In particular with PyPI as there is no review it would be fairly easy to make a package just containing (a large amount of) data.

Author: Done

    - In the interests of improving researcher productivity the aim will be to accommodate such requests, provided there are no outweighing security concerns associated with the package or its dependencies.
- Requests will be reviewed by the project team using the information provided by the user when making the request
- If approved, a requested package will be added to either the core or extra allowlist (as appropriate)

(package_inclusion_criteria)=

### Criteria for inclusion in core

Only software that is considered broadly useful to a cross section of researchers should be included in core.

To meet this condition, a package should:

- implement at least one generic (i.e. not domain-specific) statistical algorithm or method, or
- provide support for a cross-cutting analysis technique (e.g. geospatial data analysis, NLP), or
- facilitate data science or software development best practices (e.g. for robustness, correctness, reproducibility), or
- enhance the presentational features of the programming language (e.g. for producing plots, notebooks, articles, websites), or

> - enhance the usability of the programming language or development environment (e.g. RStudio, PyCharm).

Reviewer: ```suggestion
- enhance the usability of the programming language or development environment (e.g. RStudio, PyCharm)
```

# Software approval policy

To maximise the usability of the secure research environments, we pre-install certain software packages that are deemed broadly useful to a cross section of researchers, thus making them available to all users.

Other software packages which are only of interest to a subset of users can be made available for download from remote package repositories.
Currently, we support PyPI (Python) and CRAN (R) as remote repositories.

> Only a subset of packages that have been vetted to mitigate the risk of introducing malicious or unsound software into the secure environment are made available in this way.

Reviewer: Doesn't this only apply to Tier 3 +

Author: Done

## Background

Given the safeguards afforded by the safe haven environment, and the separation of responsibilities between its constituent resources, the level of risk associated with the introduction of third party software packages is considered low.
Moreover, access to the environment is carefully controlled and there is a presumption of trust in the individual researchers to whom access is granted.

Nevertheless, the introduction of any software into the safe haven must be considered against the potential risks of:

- approved users having access to data to which they shouldn't (e.g. from data mixing)
- unapproved users having access to data (e.g. from a data breach)
- poisoning of data and/or outputs
- resource misuse (allocation of computational resources for unintended or wasteful purposes).

Such risks may originate unwittingly, from a user who wants to "just get the job done", or from a user, network team member or administrator acting maliciously.

Specific risks which this policy aims to mitigate include:

- package name squatting (allowlisting a similarly-named package instead of the intended one)

> - privilege escalation attacks (enabling a user to gain elevated access permissions).

Reviewer: ```suggestion
- privilege escalation attacks (enabling a user to gain elevated access permissions)
```

To be consistent.

(package_inclusion_policy)=

## Policy

> - For each supported programming language, three package lists will be maintained:

Reviewer: ```suggestion
- For each supported repository, three package lists will be maintained:
```

  - a core allowlist of broadly useful packages that should be pre-installed in each environment
  - an extra allowlist of packages that may be useful for specific projects
  - an expanded list to be made available from the package repositories consisting of the core and extra packages plus their dependencies
- Users may request to add packages to these allowlists via the {ref}`package request procedure <package_request_procedure>`.
  - In the interests of improving researcher productivity the aim will be to accommodate such requests, provided there are no outweighing security concerns associated with the package or its dependencies.
- Requests will be reviewed by the project team using the information provided by the user when making the request

> - If approved, a requested package will be added to either the core or extra allowlist (as appropriate)

Reviewer: Again, do we really want to do this? It sounds like we are taking on quite a bit of burden in reviewing and adding packages.

Does this work for the end users? It would be faster for the admins of their instance to update the allowlists than to go through review, merge, release, update.

Author: Tier-3 allowlists are currently SHM-wide. Allowing all projects access to a particular package is a reasonably big decision. Any package that convinces the DSH administrators that it should be allowed to all projects is presumably a reasonable candidate to be in our "default" allowlist.

Running a default allowlist is also a usability feature - we're not saying "work out what's allowed" to each deployment, we're saying "here's a default, but use your own if you prefer".

I agree, it would definitely be faster to manually update the allowlist, but that means that the work that's been put into evaluating the package doesn't come back to the upstream project.

Reviewer: Oh, I had forgotten it was per SHM. Have we thought about when we might want to have different allow lists per project?

In that case, anything that gets approved is much closer to being one of the pre-installed packages than I was imagining.

> "here's a default, but use your own if you prefer"

I suppose I'm a little nervous about the risk of us getting it wrong. It feels a touch more like a guarantee than other bits of policy where we just say "this is what _we_ do, but you need to think about your own situation".

Author: That's a good point but we *do* have an existing Tier-3 list which people who run an out-of-the-box deployment will be using. We might not be comfortable in curated a set of packages but the fact is that we already are. Maybe our focus should be on switching to a community-developed list instead?

Reviewer: Sounds good. In terms of actions, shall we

1. Add this to the project board as an item for RSE-TR-WG?
2. See if this fits into any part of SATRE?

Author: Definitely (1) and we can ask about (2) but that might be better covered by one of the other DARE-funded projects.

- implement at least one generic (i.e. not domain-specific) statistical algorithm or method, or
- provide support for a cross-cutting analysis technique (e.g. geospatial data analysis, NLP), or
- facilitate data science or software development best practices (e.g. for robustness, correctness, reproducibility), or
- enhance the presentational features of the programming language (e.g. for producing plots, notebooks, articles, websites), or
- enhance the usability of the programming language or development environment (e.g. RStudio, PyCharm).

(package_request_procedure)=

## Package request/review procedure

1. A user requests a package by opening a `Software package request` issue on the Data Safe Haven GitHub repository, including responses to the following questions:

- Is this package the mostly widely supported for the intended purpose?
- What will you be able to do with this package that you can't currently do? What alternatives are there?
- What risks to data integrity/security might arise from including this package or its dependencies?

2. A member of the project team reviews the request according to the terms of the {ref}`package_inclusion_policy`.

> 3. The reviewer adds their decision (accept/reject) to the issue and notifies the user who made the request.

Reviewer 2: If I was a reviewer, how would I know how to make this decision? Is it possible to expand some guidance here - for example, in Daniel from Edon's case, would the reviewer just google "R Arrow" and check it's a real package and somehow figure out if it's something without known security vulnerabilities?

Author: The issue template asks them to answer the following questions:

- Package name
- Package version (if different from latest)
- Package repository (e.g. CRAN, PyPI)
- Number of authors/contributors to the package codebase
- Any existing versions that should not be used (linking to publicly-accessible CVE databases if relevant)
- Download statistics (recent and longer-term, for both current and previous versions)
- List of packages that this package depends on
- What will you be able to do with this package that you can't currently do?
- Is this the most widely supported package for the intended purpose? What alternatives are there?
- What risks to data integrity/security might arise from including this package or its dependencies?

Reviewer 2: Makes sense, so the burden of checking risks falls to the requester

Author: The burden of explaining why the risks are manageable. We still don't have a well-defined way to make a decision based on these answers though :)
